// @nomicfoundation/hardhat-mocha/dist/src/performance.js
import { performance } from "node:perf_hooks";
/**
 * Creates a new performance tracker for measuring an operation and
 * its sub-phases.
 *
 * @param scope - A namespace prefix for performance marks (e.g.
 *   "hardhat:mocha:performance").
 * @param trackerName - The display name for the overarching measure
 *   (e.g. "Mocha test task").
 * @returns A new {@link PerformanceTracker} instance.
 */
export function createPerformanceTracker(scope, trackerName) {
    return new PerformanceTrackerImpl(scope, trackerName);
}
class PerformanceTrackerImpl {
    #scope;
    #trackerName;
    #seenPhases = [];
    constructor(scope, trackerName) {
        this.#scope = scope;
        this.#trackerName = trackerName;
    }
    start() {
        performance.mark(`${this.#markName(this.#trackerName)}:start`);
    }
    end() {
        const markName = this.#markName(this.#trackerName);
        performance.mark(`${markName}:end`);
        performance.measure(this.#trackerName, `${markName}:start`, `${markName}:end`);
    }
    startPhase(phase) {
        performance.mark(`${this.#markName(phase)}:start`);
    }
    endPhase(phase) {
        const markName = this.#markName(phase);
        performance.mark(`${markName}:end`);
        performance.measure(phase, `${markName}:start`, `${markName}:end`);
        this.#seenPhases.push(phase);
    }
    logInto(debugLog) {
        const measures = performance
            .getEntriesByType("measure")
            .filter((m) => m.name === this.#trackerName || this.#seenPhases.includes(m.name));
        const longestMeasureName = measures
            .map(({ name }) => name === this.#trackerName ? name.length + 2 : name.length)
            .reduce((acc, v) => Math.max(acc, v), 0);
        for (const m of measures) {
            const indent = m.name === this.#trackerName ? false : true;
            const timingIndent = " ".repeat(longestMeasureName - m.name.length + (indent ? 1 : 3));
            debugLog(`${indent ? "  " : ""}${m.name}:${timingIndent}${m.duration.toFixed(2)}ms`);
        }
    }
    clear() {
        const markName = this.#markName(this.#trackerName);
        performance.clearMarks(`${markName}:start`);
        performance.clearMarks(`${markName}:end`);
        performance.clearMeasures(this.#trackerName);
        for (const phase of this.#seenPhases) {
            const phaseMarkName = this.#markName(phase);
            performance.clearMarks(`${phaseMarkName}:start`);
            performance.clearMarks(`${phaseMarkName}:end`);
            performance.clearMeasures(phase);
        }
        this.#seenPhases = [];
    }
    #markName(phase) {
        return `${this.#scope}:${phase}`;
    }
}
